servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletHttpRequest.java

```diff
@@ -0,0 +1,640 @@
+    }
+
+    @Override
+    public boolean isRequestedSessionIdFromCookie() {
+        return sessionIdSource == SessionTrackingMode.COOKIE;
+    }
+
+    @Override
+    public boolean isRequestedSessionIdFromURL() {
+        return isRequestedSessionIdFromUrl();
+    }
+
+    @Override
+    public boolean isRequestedSessionIdFromUrl() {
+        return sessionIdSource == SessionTrackingMode.URL;
+    }
+
+    @Override
+    public String getRequestedSessionId() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    @Override
+    @Nullable
+    public Object getAttribute(String name) {
+        requireNonNull(name, "name");
+        final Object value = getAttributeMap().get(name);
+        return value;
+    }
+
+    @Override
+    public Enumeration<String> getAttributeNames() {
+        return Collections.enumeration(ImmutableSet.copyOf(getAttributeMap().keySet()));
+    }
+
+    @Override
+    public String getCharacterEncoding() {
+        return characterEncoding;
+    }
+
+    @Override
+    public void setCharacterEncoding(String env) throws UnsupportedEncodingException {
+        throw new IllegalStateException("Can't set character encoding after request is initialized");
+    }
+
+    @Override
+    public int getContentLength() {
+        return httpRequest.content().length();
+    }
+
+    @Override
+    public long getContentLengthLong() {
+        return getContentLength();
+    }
+
+    @Override
+    @Nullable
+    public String getContentType() {
+        return httpRequest.headers().contentType().toString();
+    }
+
+    @Override
+    public DefaultServletInputStream getInputStream() throws IOException {
+        if (reader != null) {
+            throw new IllegalStateException("getReader() has already been called for this request");
+        }
+        return inputStream;
+    }
+
+    @Override
+    @Nullable
+    public String getParameter(String name) {
+        requireNonNull(name, "name");
+        return queryParams.get(name);
+    }
+
+    @Override
+    public Enumeration<String> getParameterNames() {
+        return Collections.enumeration(ImmutableSet.copyOf(queryParams.names()));
+    }
+
+    @Override
+    @Nullable
+    public String[] getParameterValues(String name) {
+        requireNonNull(name, "name");
+        if (queryParams.getAll(name) == null) {
+            return null;
+        }
+        return queryParams.getAll(name).toArray(new String[0]);
+    }
+
+    @Override
+    public Map<String, String[]> getParameterMap() {
+        return parameters;
+    }
+
+    @Override
+    public String getProtocol() {
+        return serviceRequestContext.sessionProtocol().uriText();
+    }
+
+    @Override
+    @Nullable
+    public String getScheme() {
+        return httpRequest.scheme();
+    }
+
+    @Override
+    @Nullable
+    public String getServerName() {
+        return ((InetSocketAddress) serviceRequestContext.remoteAddress()).getAddress().getHostAddress();
```

Comment:
We can do `serviceRequestContext.config().virtualHost().defaultHostname()`
and this is not nullable.